Source/options.h

```diff
@@ -0,0 +1,230 @@
+
+DEVILUTION_BEGIN_NAMESPACE
+
+/**
+ * @brief Base non-templated class for all options. Provides saving and loading functionality and a way to group every
+ * type of option together.
+*/
+class OptionBase {
+public:
+	/**
+	 * @brief Loads data from the ini into this instance.
+	 * @param sectionName The name of the section that contains the data for this option.
+	*/
+	virtual void Load(const std::string& sectionName) = 0;
+
+	/**
+	 * @brief Saves the data contained in this instance to the appropriate section in the ini.
+	 * @param sectionName The name of the section where the data from this instance will be saved to.
+	*/
+	virtual void Save(const std::string& sectionName) = 0;
+};
+
+/**
+ * @brief Base templated class for all options. Stores information about a given option, including its current value,
+ * default value and the name of the key inside the ini file that corresponds to it.
+ * @tparam T The underlying type of the setting.
+*/
+template <typename T>
+class Option : public OptionBase {
+public:
+	/**
+	 * @brief Initializes a new instance of the Option<T> with the specified display name and default value.
+	 * @param displayName The name to use when saving and loading the option to ini.
+	 * @param defaultValue The default value to use when the option does not yet exist.
+	*/
+	Option(const std::string& displayName, const T& defaultValue)
+	    : m_value(0)
+	    , m_displayName(displayName)
+	    , m_defaultValue(defaultValue)
+	{
+	}
+
+	/**
+	 * @brief Extracts the underlying value from this option instance.
+	 * @return The underlying value of the option.
+	*/
+	const T& operator*() const
+	{
+		return m_value;
+	}
+
+protected:
+	T m_value;
+	std::string m_displayName;
+	T m_defaultValue;
+};
+
+/**
+ * @brief A class that models a boolean option.
+*/
+class BooleanOption final : public Option<bool> {
+	using Option<bool>::Option;
+
+public:
+	void Load(const std::string& sectionName);
+	void Save(const std::string& sectionName);
+
+	/**
+	 * @brief Sets the underlying value of this option instance.
+	 * @param value The new value to be set as the current value of this option.
+	 * @return This instance with the new value set.
+	*/
+	BooleanOption& operator=(bool value);
+};
+
+/**
+ * @brief A class that models a group of options persisted in an ini section.
+*/
+class OptionGroup {
+public:
+	/**
+	 * @brief Initializes a new instance of the option group with the specified section name.
+	 * @param sectionName The name of the ini section used to read and write values.
+	*/
+	OptionGroup(const std::string& sectionName);
+
+	/**
+	 * @brief Loads the data from the ini file into this option group.
+	*/
+	void Load();
+
+	/**
+	 * @brief Saves the data stored in this option group to the ini file.
+	*/
+	void Save();
+
+protected:
+	void AddOption(OptionBase&);
+	void AddOptionGroup(OptionGroup&);
+
+private:
+	std::vector<OptionBase*> m_options;
+	std::vector<OptionGroup*> m_optionGroups;

```

Comment:
If this class owns the options, these vectors should be `std::vector<std::unique_ptr<OptionBase>>` (and same for option groups)

Reply:
Can you give me a bit of context? Please understand that I'm mostly not used to C++. Why is the current approach less than ideal?

Reply:
As discussed in chat, this is not needed